File: scala3/src/test/scala/io/github/sps23/typeclasses/ShowTest.scala
package io.github.sps23.typeclasses

import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers

class ShowTest extends AnyFunSuite with Matchers:

  import Show.*
  import Show.given

  test("Show[Int] should format integers correctly") {
    Show.showValue(42) shouldBe "Int(42)"
    Show.showValue(-5) shouldBe "Int(-5)"
  }

  test("Show[String] should format strings correctly") {
    Show.showValue("hello") shouldBe "\"hello\""
    Show.showValue("") shouldBe "\"\""
  }

  test("Show[Boolean] should format booleans correctly") {
    Show.showValue(true) shouldBe "True"
    Show.showValue(false) shouldBe "False"
  }

  test("Show[Double] should format doubles correctly") {
    Show.showValue(3.14) shouldBe "Double(3.14)"
  }

  test("Show[List[Int]] should format lists correctly") {
    Show.showValue(List(1, 2, 3)) shouldBe "List(Int(1), Int(2), Int(3))"
    Show.showValue(List.empty[Int]) shouldBe "List()"
  }

  test("Show[Option[Int]] should format options correctly") {
    Show.showValue(Some(42): Option[Int]) shouldBe "Some(Int(42))"
    Show.showValue(None: Option[Int]) shouldBe "None"
  }

  test("Show[Map[String, Int]] should format maps correctly") {
    val result = Show.showValue(Map("a" -> 1))
    result shouldBe "Map(\"a\" -> Int(1))"
  }

  test("Extension method .show should work") {
    42.show shouldBe "Int(42)"
    "hello".show shouldBe "\"hello\""
    List(1, 2, 3).show shouldBe "List(Int(1), Int(2), Int(3))"
  }

  test("Custom Show instance should work") {
    case class Person(name: String, age: Int)

    given Show[Person] with
      def show(p: Person): String = s"Person(${p.name}, ${p.age})"

    Show.showValue(Person("Alice", 30)) shouldBe "Person(Alice, 30)"
    Person("Bob", 35).show shouldBe "Person(Bob, 35)"
  }

  test("Nested collections should work") {
    val nested = List(List(1, 2), List(3, 4))
    nested.show shouldBe "List(List(Int(1), Int(2)), List(Int(3), Int(4)))"
  }

  test("Context bounds should work") {
    def printValue[A: Show](a: A): String = Show.showValue(a)

    printValue(42) shouldBe "Int(42)"
    printValue("test") shouldBe "\"test\""
  }

  test("Summoning instances should work") {
    val intShow = summon[Show[Int]]
    intShow.show(99) shouldBe "Int(99)"

    val stringShow = Show[String]
    stringShow.show("scala") shouldBe "\"scala\""
  }

  test("List of custom types should work with automatic derivation") {
    case class Person(name: String, age: Int)

    given Show[Person] with
      def show(p: Person): String = s"Person(${p.name}, ${p.age})"

    val people = List(Person("Alice", 30), Person("Bob", 35))
    people.show shouldBe "List(Person(Alice, 30), Person(Bob, 35))"
  }

File: java21/src/test/java/io/github/sps23/typeclasses/ShowTest.java
package io.github.sps23.typeclasses;

import static org.junit.jupiter.api.Assertions.*;

import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

@DisplayName("Show Typeclass Tests")
class ShowTest {

    @Nested
    @DisplayName("Version 1: Basic Implementation")
    class BasicImplementationTests {

        @Test
        @DisplayName("Should show integers correctly")
        void testIntegerShow() {
            Show<Integer> intShow = Show.forInteger();
            assertEquals("Int(42)", intShow.show(42));
            assertEquals("Int(-5)", intShow.show(-5));
        }

        @Test
        @DisplayName("Should show strings correctly")
        void testStringShow() {
            Show<String> stringShow = Show.forString();
            assertEquals("\"hello\"", stringShow.show("hello"));
            assertEquals("\"\"", stringShow.show(""));
        }

        @Test
        @DisplayName("Should show booleans correctly")
        void testBooleanShow() {
            Show<Boolean> boolShow = Show.forBoolean();
            assertEquals("True", boolShow.show(true));
            assertEquals("False", boolShow.show(false));
        }

        @Test
        @DisplayName("Should show lists correctly")
        void testListShow() {
            Show<List<Integer>> listShow = Show.forList(Show.forInteger());
            assertEquals("List(Int(1), Int(2), Int(3))", listShow.show(List.of(1, 2, 3)));
            assertEquals("List()", listShow.show(List.of()));
        }

        @Test
        @DisplayName("Should show maps correctly")
        void testMapShow() {
            Show<Map<String, Integer>> mapShow = Show.forMap(Show.forString(), Show.forInteger());
            String result = mapShow.show(Map.of("a", 1));
            assertEquals("Map(\"a\" -> Int(1))", result);
        }

        @Test
        @DisplayName("Should handle null values")
        void testNullHandling() {
            Show<String> show = Show.forAny();
            assertEquals("null", show.show(null));
        }
    }

    @Nested
    @DisplayName("Version 2: Registry-Based Lookup")
    class RegistryTests {

        @Test
        @DisplayName("Should retrieve registered instances")
        void testRegistryLookup() {
            Show<Integer> intShow = Show.Registry.summon(Integer.class);
            assertEquals("Int(42)", intShow.show(42));

            Show<String> stringShow = Show.Registry.summon(String.class);
            assertEquals("\"hello\"", stringShow.show("hello"));
        }

        @Test
        @DisplayName("Should register custom instances")
        void testCustomRegistration() {
            record Person(String name, int age) {
            }

            Show.Registry.register(Person.class, p -> "Person(" + p.name() + ", " + p.age() + ")");

            Show<Person> personShow = Show.Registry.summon(Person.class);
            assertEquals("Person(Alice, 30)", personShow.show(new Person("Alice", 30)));
        }

        @Test
        @DisplayName("Should fallback to toString for unregistered types")
        void testFallbackBehavior() {
            record Unknown(int x) {
            }

            Show<Unknown> show = Show.Registry.summon(Unknown.class);
            Unknown u = new Unknown(42);
            assertEquals(u.toString(), show.show(u));
        }
    }

    @Nested
    @DisplayName("Version 4: Extension Methods")
    class ShowableTests {

        @Test
        @DisplayName("Should wrap and show values")
        void testShowableBasic() {
            String result = Showable.of(42, Show.forInteger()).show();
            assertEquals("Int(42)", result);
        }

        @Test
        @DisplayName("Should support mapping")
        void testShowableMap() {
            String result = Showable.of(21, Show.forInteger()).map(i -> i * 2, Show.forInteger())
                    .show();
            assertEquals("Int(42)", result);
        }

        @Test
        @DisplayName("Should work with registry lookup")
        void testShowableWithRegistry() {
            String result = Showable.of(42, Integer.class).show();
            assertEquals("Int(42)", result);
        }
    }

    @Nested
    @DisplayName("Version 5: Auto-Derivation")
    class DerivationTests {

        @Test
        @DisplayName("Should derive Show for simple records")
        void testSimpleRecordDerivation() {
            record Person(String name, int age) {
            }

            Show<Person> personShow = ShowDerivation.deriveRecord();
            Person alice = new Person("Alice", 30);

            String result = personShow.show(alice);
            assertTrue(result.contains("Person"));
            assertTrue(result.contains("name = \"Alice\""));
            assertTrue(result.contains("age = 30"));
        }

        @Test
        @DisplayName("Should derive Show for nested records")
        void testNestedRecordDerivation() {
            record Address(String street, String city) {
            }
            record Person(String name, Address address) {
            }

            Show<Person> personShow = ShowDerivation.deriveRecord();
            Person bob = new Person("Bob", new Address("123 Main St", "Springfield"));

            String result = personShow.show(bob);
            assertTrue(result.contains("Person"));
            assertTrue(result.contains("name = \"Bob\""));
            assertTrue(result.contains("Address"));
            assertTrue(result.contains("street = \"123 Main St\""));
            assertTrue(result.contains("city = \"Springfield\""));
        }

        @Test
        @DisplayName("Should handle null in record fields")
        void testNullInRecordFields() {
            record Person(String name, Integer age) {
            }

            Show<Person> personShow = ShowDerivation.deriveRecord();
            Person person = new Person(null, null);

            String result = personShow.show(person);
            assertTrue(result.contains("name = null"));
            assertTrue(result.contains("age = null"));
        }
    }
}
